Common/Securities/CurrencyConversion/SecurityCurrencyConversion.cs

```diff
@@ -0,0 +1,268 @@
+                }
+
+                if (step.Inverted)
+                {
+                    newConversionRate /= price;
+                }
+                else
+                {
+                    newConversionRate *= price;
+                }
+            });
+
+            ConversionRate = newConversionRate;
+            return ConversionRate;
+        }
+
+        /// <summary>
+        /// Finds a conversion between two currencies by looking through all available 1 and 2-leg options
+        /// </summary>
+        /// <param name="sourceCurrency">The currency to convert from</param>
+        /// <param name="destinationCurrency">The currency to convert to</param>
+        /// <param name="existingSecurities">The securities which are already added to the algorithm</param>
+        /// <param name="potentialSymbols">The symbols to consider, may overlap with existingSecurities</param>
+        /// <param name="makeNewSecurity">The function to call when a symbol becomes part of the conversion, must return the security that will provide price data about the symbol</param>
+        /// <returns>A new <see cref="SecurityCurrencyConversion"/> instance representing the conversion from sourceCurrency to destinationCurrency</returns>
+        /// <exception cref="ArgumentException">Thrown when no conversion from sourceCurrency to destinationCurrency can be found</exception>
+        public static SecurityCurrencyConversion LinearSearch(
+            string sourceCurrency,
+            string destinationCurrency,
+            IList<Security> existingSecurities,
+            IEnumerable<Symbol> potentialSymbols,
+            Func<Symbol, Security> makeNewSecurity)
+        {
+            var allSymbols = existingSecurities.Select(sec => sec.Symbol).Concat(potentialSymbols)
+                .Where(x => x.SecurityType == SecurityType.Crypto || x.Value.Length == 6)
+                .ToList();
+
+            var securitiesBySymbol = existingSecurities.Aggregate(new Dictionary<Symbol, Security>(),
+                (mapping, security) =>
+                {
+                    if (!mapping.ContainsKey(security.Symbol))
+                    {
+                        mapping[security.Symbol] = security;
+                    }
+
+                    return mapping;
+                });
+
+            // Search for 1 leg conversions
+            foreach (var potentialConversionRateSymbol in allSymbols)
+            {
+                if (!potentialConversionRateSymbol.PairContainsCurrency(sourceCurrency))
+                {
+                    continue;
+                }
+
+                if (potentialConversionRateSymbol.CurrencyPairDual(sourceCurrency) != destinationCurrency)
+                {
+                    continue;
+                }
+
+                var steps = new List<Step>(1);
+
+                var inverted = potentialConversionRateSymbol.ComparePair(sourceCurrency, destinationCurrency) ==
```

Comment:
Seems we could replace `PairContainsCurrency/CurrencyPairDual & ComparePair` for a single call, would be nice if the 2step bellow and 1 step here followed the same pattern to determine if matched, inverted or not